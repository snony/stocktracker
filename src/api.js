const host = 'https://api.iextrading.com/1.0/'

const fetchAndUnpack = url => fetch(url).then(data => data.json())

export const getHistory = (symbol, dateFilter = 'ytd', priceFilter = 'close') => {
  const url = `${host}stock/${symbol}/chart/${dateFilter}?filter=date,${priceFilter}`
  return fetchAndUnpack(url)
}
export const getRefData = () => {
  const url = `${host}ref-data/symbols?filter=name,symbol`
  return fetchAndUnpack(url)
}

export const getNews = symbol => {
  const url = `${host}stock/${symbol}/news/last/5`
  return fetchAndUnpack(url)
}

export const getOverview = symbol => {
  const url = `${host}stock/${symbol}/company`
  return fetchAndUnpack(url)
}

export const getPeers = symbol => {
  const url = `${host}stock/${symbol}/peers`
  return fetchAndUnpack(url)
}

export const getStats = symbol => {
  const url = `${host}stock/${symbol}/stats`
  return fetchAndUnpack(url)
}

export const getPrevious = symbol => {
  const url = `${host}stock/${symbol}/previous`
  return fetchAndUnpack(url)
}
export const getOhlc = symbol => {
  const url = `${host}stock/${symbol}/ohlc`
  return fetchAndUnpack(url)
}
export const getQuote = symbol => {
  const url = `${host}stock/${symbol}/quote`
  return fetchAndUnpack(url)
}

const keyStatsServices = [getStats, getPrevious, getOhlc, getQuote]

const extractData = ([stats, previous, ohlc, quote]) => ({
  earningsPerShare: stats.latestEPS,
  dividendYield: stats.dividendYield,
  previousClose: previous.close,
  open: ohlc.open.price,
  close: ohlc.close.price,
  dayHigh: ohlc.high,
  dayLow: ohlc.low,
  volume: quote.latestVolume,
  avgTotalVolume: quote.avgTotalVolume,
  peRatio: quote.peRatio,
  marketCap: quote.marketCap,
  week52High: quote.week52High,
  week52Low: quote.week52Low
})

export const getKeyStats = symbol => {
  return Promise.all(keyStatsServices.map(service => service(symbol))).then(extractData)
}

const companyInfoServices = [getHistory, getNews, getOverview, getPeers, getKeyStats]
const extractCompanyInfo = ([history, news, overview, peers, keystats]) => ({
  history,
  news,
  overview,
  peers,
  keystats
})

export const getCompanyInfo = symbol => {
  return Promise.all(companyInfoServices.map(service => service(symbol))).then(extractCompanyInfo)
}
