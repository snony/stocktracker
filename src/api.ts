const host = 'https://api.iextrading.com/1.0'

const fetchAndUnpack = async <T>(url: string): Promise<T> => {
  const data = await fetch(url)
  return await data.json()
}

export const getCompanySymbols = async () => {
  const url = `${host}/ref-data/symbols?filter=name,symbol`
  return await fetchAndUnpack(url)
}

export const getHistory = async (
  symbol: string,
  dateFilter: string = 'ytd',
  priceFilter: string = 'close'
) => {
  const url = `${host}/stock/${symbol}/chart/${dateFilter}?filter=date,${priceFilter}`
  return await fetchAndUnpack(url)
}

export const getNews = async (symbol: string) => {
  const url = `${host}/stock/${symbol}/news/last/5`
  return await fetchAndUnpack(url)
}

export const getOverview = async (symbol: string) => {
  const url = `${host}/stock/${symbol}/company`
  return await fetchAndUnpack(url)
}

export const getPeers = async (symbol: string) => {
  const url = `${host}/stock/${symbol}/peers`
  return await fetchAndUnpack(url)
}

interface StockData {
  [key: string]: number
}

interface KeyStatsData {
  quote: StockData
  stats: StockData
}

export const extractData = ({ quote, stats }: KeyStatsData) => ({
  earningsPerShare: stats.latestEPS,
  dividendYield: stats.dividendYield,
  previousClose: quote.previousClose,
  open: quote.open,
  dayHigh: quote.high,
  dayLow: quote.low,
  volume: quote.latestVolume,
  avgTotalVolume: quote.avgTotalVolume,
  peRatio: quote.peRatio,
  marketCap: quote.marketCap,
  week52High: quote.week52High,
  week52Low: quote.week52Low
})

export const getKeyStats = async (symbol: string) => {
  const url = `${host}/stock/${symbol}/batch?types=quote,stats`
  const keystats = await fetchAndUnpack<KeyStatsData>(url)
  return extractData(keystats)
}

export default {
  getCompanySymbols,
  getHistory,
  getNews,
  getKeyStats,
  getOverview,
  getPeers
}
