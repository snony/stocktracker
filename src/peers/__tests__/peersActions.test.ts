import { mockApi } from '__mock__/api.mock'
import { mockGlobalState } from '__mock__/globalstate.mock'
import { generateMockStore } from '__mock__/mockStore.mock'
import {
  getPeersData,
  PEERS_RECEIVED_ACTION,
  PeersReceivedAction,
  peersReceivedAction
} from 'peers/peersActions'
import { MockStore } from 'redux-mock-store'

describe('tests for peers action', () => {
  let store: MockStore<{}>

  const mockSymbol = 'aapl'
  const mockData = mockGlobalState.peers
  const expectedAction: PeersReceivedAction = {
    type: PEERS_RECEIVED_ACTION,
    peers: mockData
  }

  beforeEach(() => {
    store = generateMockStore(mockGlobalState, mockApi)
    store.clearActions()
  })

  describe('test for receive action', () => {
    it('should create new PEERS_RECEIVED_ACTION object', () => {
      const action = peersReceivedAction(mockData)
      expect(action).toEqual(expectedAction)
    })
  })

  describe('test for async thunk action', () => {
    it('should dispatch PEERS_RECEIVED_ACTION when the fetch is successful', async () => {
      const mockDispatch = store.dispatch
      const mockStoreActions = store.getActions()

      await mockDispatch<any>(getPeersData(mockSymbol))

      expect(mockStoreActions[0]).toEqual(expectedAction)
    })
  })
})
