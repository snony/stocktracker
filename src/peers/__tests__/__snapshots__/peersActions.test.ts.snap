// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test for peers action should create new PEERS_RECEIVED_ACTION object 1`] = `
Array [
  Array [
    Object {
      "peers": Array [
        "peers1",
        "peers2",
        "peers3",
        "peers4",
      ],
      "type": "PEERS_RECEIVED_ACTION",
    },
  ],
]
`;
